gcli/src/meta/executor.rs

```diff
@@ -0,0 +1,228 @@
+// This file is part of Gear.
+//
+// Copyright (C) 2021-2023 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! The WASM executor in this module is just for parsing the state types
+//! of gear programs, some of the host functions are missing logics that
+//! is because they are for the on-chain environment data.
+
+use anyhow::{anyhow, Result};
+use wasmi::{AsContextMut, Engine, Extern, Linker, Memory, MemoryType, Module, Store};
+
+const PAGE_STORAGE_PREFIX: [u8; 32] = *b"gcligcligcligcligcligcligcligcli";
+
+/// HostState for the WASM executor
+#[derive(Default)]
+pub struct HostState {
+    pub msg: Vec<u8>,
+}
+
+/// Executes the WASM code.
+pub fn execute(wasm: &[u8], method: &str) -> Result<Vec<u8>> {
+    assert!(gear_lazy_pages_interface::try_to_enable_lazy_pages(
+        PAGE_STORAGE_PREFIX
+    ));
+
+    let engine = Engine::default();
+    let module = Module::new(&engine, wasm).unwrap();
+
+    let mut store = Store::new(&engine, HostState::default());
+    let mut linker = <Linker<HostState>>::new();
+
+    // Execution environment
+    {
+        let memory = Memory::new(store.as_context_mut(), MemoryType::new(256, None)).unwrap();
+        linker.define("env", "memory", Extern::Memory(memory))?;
+        linker.define("env", "gr_read", funcs::gr_read(&mut store, memory))?;
+        linker.define("env", "alloc", funcs::alloc(&mut store, memory))?;
+        linker.define("env", "free", funcs::free(&mut store))?;
+        linker.define("env", "gr_size", funcs::gr_size(&mut store, memory))?;
+        linker.define("env", "gr_reply", funcs::gr_reply(&mut store, memory))?;
```

Comment:
gr_debug may appear as well

Reply:
and gr_reply_push + gr_reply_commit

Reply:
https://github.com/gear-tech/gear/pull/3366/commits/18cdf6bb414348bf9cdf3708b16567e1e8597c19